mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,929 @@
+    #[snafu(display("MetaDataIndex address is not set for child {}", name))]
+    IndexAddressNotSet { name: String },
+    #[snafu(display("Missing partition: {}", name))]
+    MissingPartition { name: String },
+}
+
+#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct MetaDataChildEntry {
+    pub guid: Guid,
+    pub state: u16,
+}
+
+impl MetaDataChildEntry {
+    /// Calculate checksum over the child entries
+    pub fn checksum(children: &[MetaDataChildEntry]) -> Result<u32, Error> {
+        let mut digest = crc32::Digest::new(crc32::IEEE);
+        let count = children.len() as u32;
+        for child in children {
+            digest.write(&serialize(child)?);
+        }
+        if count < MetaDataObject::MAX_CHILD_ENTRIES {
+            let pad = serialize(&MetaDataChildEntry::default())?;
+            for _ in count .. MetaDataObject::MAX_CHILD_ENTRIES {
+                digest.write(&pad);
+            }
+        }
+        Ok(digest.sum32())
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+struct MetaDataHeader {
+    signature: [u8; 8],
+    self_checksum: u32,
+    generation: u64,
+    timestamp: u128,
+    table_crc: u32,
+    count: u32,
+}
+
+impl MetaDataHeader {
+    /// Checksum the header with the checksum field itself set to 0
+    fn checksum(&mut self) -> Result<u32, Error> {
+        self.self_checksum = 0;
+        self.self_checksum = crc32::checksum_ieee(&serialize(&self)?);
+        Ok(self.self_checksum)
+    }
 }
 
-#[derive(Debug, Deserialize, PartialEq, Default, Serialize, Copy, Clone)]
-pub struct MetaDataHeader {
-    /// Signature identifying this as a MetaDataHeader object
+#[derive(Debug, PartialEq)]
+pub struct MetaDataObject {
     pub signature: [u8; 8],
-    /// Size of this header in bytes
-    pub header_size: u32,
-    /// CRC-32 checksum of this header
     pub self_checksum: u32,
-    /// Current object generation counter
     pub generation: u64,
-    /// Absolute location (LBA) of this header on disk
-    pub self_lba: u64,
-    /// Offset of start of index table relative to self_lba
-    pub index_start: u64,
-    /// Number of valid entries in the index
-    pub used_entries: u32,
-    /// Maximum number of entries that the index can contain
-    pub max_entries: u32,
-    /// Size of an index entry in bytes
-    pub entry_size: u32,
-    /// CRC-32 checksum of the index table
-    pub index_checksum: u32,
-    /// Offset of first usable data block relative to self_lba
-    pub data_start: u64,
-    /// Offset of last usable data block relative to self_lba
-    pub data_end: u64,
+    pub timestamp: u128,
+    pub table_crc: u32,
+    pub children: Vec<MetaDataChildEntry>,
 }
 
-impl MetaDataHeader {
-    pub const METADATA_HEADER_SIZE: u32 = 72;
-    pub const MAX_INDEX_ENTRIES: u32 = 32;
-    pub const INDEX_ENTRY_SIZE: u32 = 44;
+impl MetaDataObject {
+    /// Observe that sizeof(MetaDataHeader) + 26 * sizeof(MetaDataChildEntry) is
```

Comment:
You might be able to guarantee this with https://docs.rs/static_assertions/1.1.0/static_assertions/macro.const_assert.html